File: src/cluster_kmeans.cpp
#include <emscripten/bind.h>

#include <algorithm>
#include <memory>
#include <cstdint>
#include <cstddef>

#include "utils.h"

#include "kmeans/kmeans.hpp"

class ClusterKmeansResult {
private:
    kmeans::Results<std::int32_t, std::int32_t, double> my_store;

public:
    ClusterKmeansResult(kmeans::Results<std::int32_t, std::int32_t, double> s) : my_store(std::move(s)) {}

    const auto& store() const {
        return my_store;
    }

public:
    JsFakeInt num_obs() const {
        return int2js(my_store.clusters.size());
    }

    JsFakeInt num_clusters() const {
        return int2js(my_store.details.sizes.size());
    }

    emscripten::val clusters() const {
        const auto& c = my_store.clusters;
        return emscripten::val(emscripten::typed_memory_view(c.size(), c.data()));
    }

    emscripten::val cluster_sizes() const {
        const auto& s = my_store.details.sizes;
        return emscripten::val(emscripten::typed_memory_view(s.size(), s.data()));
    }

    JsFakeInt iterations() const {
        return int2js(my_store.details.iterations);
    }

    JsFakeInt status() const {
        return int2js(my_store.details.status);
    }

    emscripten::val centers() const {
        const auto& s = my_store.centers;
        return emscripten::val(emscripten::typed_memory_view(s.size(), s.data()));
    }
};

ClusterKmeansResult cluster_kmeans(
    JsFakeInt mat_raw,
    JsFakeInt nr_raw,
    JsFakeInt nc_raw,
    JsFakeInt k_raw,
    std::string init_method,
    JsFakeInt init_seed_raw,
    double init_varpart_size_adjust,
    double init_varpart_optimized,
    std::string refine_method,
    JsFakeInt refine_lloyd_iterations_raw,
    JsFakeInt refine_hw_iterations_raw,
    JsFakeInt nthreads_raw
) {
    kmeans::SimpleMatrix<std::int32_t, double> smat(
        js2int<std::size_t>(nr_raw),
        js2int<std::int32_t>(nc_raw),
        reinterpret_cast<const double*>(js2int<std::uintptr_t>(mat_raw))
    );

    std::unique_ptr<kmeans::Initialize<std::int32_t, double, std::int32_t, double, I<decltype(smat)> > > iptr;
    if (init_method == "random") {
        auto iptr2 = new kmeans::InitializeRandom<std::int32_t, double, std::int32_t, double, I<decltype(smat)> >;
        iptr.reset(iptr2);
        iptr2->get_options().seed = js2int<std::uint64_t>(init_seed_raw);

    } else if (init_method == "kmeans++") {
        auto iptr2 = new kmeans::InitializeKmeanspp<std::int32_t, double, std::int32_t, double, I<decltype(smat)> >;
        iptr.reset(iptr2);
        iptr2->get_options().seed = js2int<std::uint64_t>(init_seed_raw);
        iptr2->get_options().num_threads = js2int<int>(nthreads_raw);

    } else if (init_method == "var-part") {
        auto iptr2 = new kmeans::InitializeVariancePartition<std::int32_t, double, std::int32_t, double, I<decltype(smat)> >;
        iptr.reset(iptr2);
        iptr2->get_options().size_adjustment = init_varpart_size_adjust;
        iptr2->get_options().optimize_partition = init_varpart_optimized;

    } else {
        throw std::runtime_error("unknown initialization method '" + init_method + "'");
    }

    std::unique_ptr<kmeans::Refine<std::int32_t, double, std::int32_t, double, I<decltype(smat)> > > rptr;
    if (refine_method == "lloyd") {
        auto rptr2 = new kmeans::RefineLloyd<std::int32_t, double, std::int32_t, double, I<decltype(smat)> >;
        rptr.reset(rptr2);
        rptr2->get_options().max_iterations = js2int<int>(refine_lloyd_iterations_raw);
        rptr2->get_options().num_threads = js2int<int>(nthreads_raw);

    } else if (refine_method == "hartigan-wong") {
        auto rptr2 = new kmeans::RefineHartiganWong<std::int32_t, double, std::int32_t, double, I<decltype(smat)> >;
        rptr.reset(rptr2);
        rptr2->get_options().max_iterations = js2int<int>(refine_hw_iterations_raw);

    } else {
        throw std::runtime_error("unknown refinement method '" + refine_method + "'");
    }

    auto output = kmeans::compute(smat, *iptr, *rptr, js2int<std::int32_t>(k_raw));
    return ClusterKmeansResult(std::move(output));
}

EMSCRIPTEN_BINDINGS(cluster_kmeans) {
    emscripten::function("cluster_kmeans", &cluster_kmeans, emscripten::return_value_policy::take_ownership());

    emscripten::class_<ClusterKmeansResult>("ClusterKmeansResult")
        .function("num_obs", &ClusterKmeansResult::num_obs, emscripten::return_value_policy::take_ownership())
        .function("num_clusters", &ClusterKmeansResult::num_clusters, emscripten::return_value_policy::take_ownership())
        .function("cluster_sizes", &ClusterKmeansResult::cluster_sizes, emscripten::return_value_policy::take_ownership())
        .function("clusters", &ClusterKmeansResult::clusters, emscripten::return_value_policy::take_ownership())
        .function("centers", &ClusterKmeansResult::centers, emscripten::return_value_policy::take_ownership())
        .function("iterations", &ClusterKmeansResult::iterations, emscripten::return_value_policy::take_ownership())
        .function("status", &ClusterKmeansResult::status, emscripten::return_value_policy::take_ownership())
        ;
}

File: js/clusterKmeans.js
import * as utils from "./utils.js";
import * as gc from "./gc.js";
import { RunPcaResults } from "./runPca.js";

/**
 * Wrapper around the k-means clustering results on the Wasm heap, produced by {@linkcode clusterKmeans}.
 * @hideconstructor
 */
export class ClusterKmeansResults {
    #id;
    #results;

    constructor(id, raw, filled = true) {
        this.#results = raw;
        this.#id = id;
    }

    /**
     * @return {number} Number of cells in the results.
     */
    numberOfCells() {
        return this.#results.num_obs();
    }

    /**
     * @return {number} Number of clusters in the results.
     */
    numberOfClusters() {
        return this.#results.num_clusters();
    }

    /**
     * @param {object} [options={}] - Optional parameters.
     * @param {boolean|string} [options.copy=true] - Whether to copy the results from the Wasm heap, see {@linkcode possibleCopy}.
     * @return {Int32Array|Int32WasmArray} Array containing the cluster assignment for each cell.
     */
    clusters({ copy = true } = {}) {
        return utils.possibleCopy(this.#results.clusters(), copy);
    }

    /**
     * @param {object} [options={}] - Optional parameters.
     * @param {boolean|string} [options.copy=true] - Whether to copy the results from the Wasm heap, see {@linkcode possibleCopy}.
     * @return {Int32Array|Int32WasmArray} Array containing the number of cells in each cluster.
     */
    sizes({ copy = true } = {}) {
        return utils.possibleCopy(this.#results.cluster_sizes(), copy);
    }

    /**
     * @param {object} [options={}] - Optional parameters.
     * @param {boolean|string} [options.copy=true] - Whether to copy the results from the Wasm heap, see {@linkcode possibleCopy}.
     * @return {Float64Array|Float64WasmArray} Array containing the cluster centers in column-major format,
     * where rows are dimensions and columns are the clusters.
     */
    centers({ copy = true } = {}) {
        return utils.possibleCopy(this.#results.centers(), copy);
    }

    /**
     * @return {number} Number of refinement iterations performed by the algorithm.
     */
    iterations() {
        return this.#results.iterations();
    }

    /**
     * @return {number} Status of the algorithm - anything other than zero usually indicates a problem with convergence.
     */
    status() {
        return this.#results.status();
    }

    /**
     * @return Frees the memory allocated on the Wasm heap for this object.
     * This invalidates this object and all references to it.
     */
    free() {
        if (this.#results !== null) {
            gc.release(this.#id);
            this.#results = null;
        }
        return;
    }
}

/**
 * Cluster cells using k-means.
 * A variety of initialization and refinement algorithms can be used here, see the [**kmeans** documentation](https://github.com/LTLA/CppKmeans) for more details.
 *
* @param {(RunPcaResults|Float64WasmArray|Array|TypedArray)} x - Numeric coordinates of each cell in the dataset.
 * For array inputs, this is expected to be in column-major format where the rows are the variables and the columns are the cells.
 * For a {@linkplain RunPcaResults} input, we extract the principal components.
 * @param {number} clusters Number of clusters to create.
 * This should not be greater than the number of cells.
 * @param {object} [options={}] - Optional parameters.
 * @param {?number} [options.numberOfDims=null] - Number of variables/dimensions per cell.
 * Only used (and required) for array-like `x`.
 * @param {?number} [options.numberOfCells=null] - Number of cells.
 * Only used (and required) for array-like `x`.
 * @param {string} [options.initMethod="pca-part"] - Initialization method.
 * Setting `"random"` will randomly select `clusters` cells as centers.
 * Setting `"kmeans++"` will use the weighted sampling approach of Arthur and Vassilvitskii (2007).
 * Setting `"var-part"` will use variance partitioning from Su and Dy (2007).
 * @param {number} [options.initSeed=5768] - Seed to use for random number generation during initialization.
 * @param {number} [options.initVarPartSizeAdjust=1] - Adjustment factor for the cluster sizes, used when `initMethod = "var-part"`.
 * Larger values (up to 1) will prioritize partitioning of clusters with more cells.
 * @param {boolean} [options.initVarPartOptimize=true] - Whether to optimize the partition at each step to minimize the sum of squares, when `initMethod = "var-part"`.
 * @param {string} [options.refineMethod="hartigan-wong"] - Refinement method.
 * This can be either `"hartigan-wong"` or `"lloyd"`.
 * @param {number} [options.refineLloydIterations=10] - Number of iterations for the Lloyd refinement algorithm.
 * @param {number} [options.refineHartiganWong=10] - Number of iterations for the Hartigan-Wong refinement algorithm.
 * @param {?number} [options.numberOfThreads=null] - Number of threads to use.
 * If `null`, defaults to {@linkcode maximumThreads}.
 *
 * @return {ClusterKmeansResults} Object containing the clustering results.
 */
export function clusterKmeans(x, clusters, { 
    numberOfDims = null, 
    numberOfCells = null, 
    initMethod = "var-part", 
    initSeed = 5768, 
    initVarPartSizeAdjust = 1, 
    initVarPartOptimize = true, 
    refineMethod = "hartigan-wong",
    refineLloydIterations = 100,
    refineHartiganWongIterations = 10,
    numberOfThreads = null 
} = {}) {
    var buffer;
    var output;
    let nthreads = utils.chooseNumberOfThreads(numberOfThreads);

    try {
        let pptr;

        if (x instanceof RunPcaResults) {
            numberOfDims = x.numberOfPCs();
            numberOfCells = x.numberOfCells();
            let pcs = x.principalComponents({ copy: false });
            pptr = pcs.byteOffset;

        } else {
            if (numberOfDims === null || numberOfCells === null) {
                throw new Error("'numberOfDims' and 'numberOfCells' must be specified when 'x' is an Array");
            }

            buffer = utils.wasmifyArray(x, "Float64WasmArray");
            if (buffer.length != numberOfDims * numberOfCells) {
                throw new Error("length of 'x' must be the product of 'numberOfDims' and 'numberOfCells'");
            }

            pptr = buffer.offset;
        }

        output = gc.call(
            module => module.cluster_kmeans(
                pptr,
                numberOfDims,
                numberOfCells,
                clusters,
                initMethod,
                initSeed,
                initVarPartSizeAdjust,
                initVarPartOptimize,
                refineMethod,
                refineLloydIterations,
                refineHartiganWongIterations,
                nthreads
            ),
            ClusterKmeansResults
        );

    } catch (e) {
        utils.free(output);
        throw e;

    } finally {
        utils.free(buffer);
    }

    return output;
}
